File: leet-code/reorganizeString.js
/**
 * @param {string} S
 * @return {string}
 */
var reorganizeString = function (S) {
  const letterDict = new Array(26).fill(0)
  const len = S.length
  for (let i = 0; i < len; i++) {
    letterDict[S[i].charCodeAt() - 97]++
  }
  const max = Math.max(...letterDict)
  if (max > Math.floor((len + 1) / 2)) {
    return ''
  }
  let even = 0
  let odd = 1
  const halfLength = Math.floor(len / 2)
  const res = []
  for (let i = 0; i <= 25; i++) {
    const c = String.fromCharCode(97 + i)
    while (letterDict[i] > 0 && letterDict[i] <= halfLength && odd < len) {
      res[odd] = c
      letterDict[i]--
      odd += 2
    }
    while (letterDict[i] > 0) {
      res[even] = c
      letterDict[i]--
      even += 2
    }
  }
  return res.join('')
}

console.log(reorganizeString('zhmyo'))

File: leet-code/reorganizeString.go
package main

import (
	"math"
	"strings"
)

func reorganizeString(S string) string {
	var dict [26]int
	l := len(S)
	res := make([]string, l+1)
	for i := range S {
		dict[S[i]-'a']++
	}

	// 找到最大值,判断能否重构
	_, max := len(dict), -math.MaxInt64
	for _, v := range dict {
		if v > max {
			max = v
		}
	}
	if max > (l+1)/2 {
		return ""
	}
	// 开始重构
	odd := 1
	even := 0
	halfC := l / 2
	for i := range dict {
		c := string(rune(i + 97))
		for dict[i] > 0 && dict[i] <= halfC && odd < l {
			res[odd] = c
			dict[i]--
			odd += 2
		}
		for dict[i] > 0 && even < l {
			res[even] = c
			dict[i]--
			even += 2
		}
	}
	return strings.Join(res, "")
}
